src/session/client-certificate.c

```diff
@@ -328,8 +377,37 @@ cockpit_session_client_certificate_map_user (const char *client_certificate_file
       return NULL;
     }
 
+  /* check if stdin is a Unix socket; then we are being called via cockpit-session@.service and need to
+   * check the cgroup of our peer */
+  struct ucred ucred;
+  int ws_cgroup_fd;
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERCRED, &ucred, &(socklen_t){sizeof ucred}) == 0)
+    {
+      debug ("unix socket mode, reading cgroup from peer pid %d", ucred.pid);
+      ws_cgroup_fd = open_proc_pid (ucred.pid);
+      int my_fd = open_proc_pid (getpid ());
+      unsigned long ws_start_time = get_proc_pid_start_time (ws_cgroup_fd);
+      unsigned long my_start_time = get_proc_pid_start_time (my_fd);
+      debug ("peer start time: %lu, my start time: %lu", ws_start_time, my_start_time);
+      close (my_fd);
+
+      /* guard against pid recycling: ws must have started earlier than ourselves */
```

Comment:
Maybe write a bit more about the attack scenario here.

Reply:
Another thing we could do is `alarm()` ourselves to prevent people from keeping long-open connections to us.  That's highly suspicious and really only useful for the attack scenario outlined in the PR.  The -ws should send the authorize message almost immediately, no?

Reply:
> Maybe write a bit more about the attack scenario here.

It's in the commit message, but I put it into the comment as well.

> alarm()

Ah, we have that on the ws side, but not in session. I'll check that tomorrow.

Reply:
I added a commit which does the alarm() thing.